core/base-service/openapi.js

```diff
@@ -332,4 +332,34 @@ function category2openapi(category, services) {
   return spec
 }
 
-export { category2openapi }
+function pathParam({
+  name,
+  example,
+  schema = { type: 'string' },
+  description,
+}) {
+  return { name, in: 'path', required: true, schema, example, description }
+}
+
+function queryParam({
+  name,
+  example,
+  schema = { type: 'string' },
+  required = false,
+  description,
+}) {
```

Comment:
I think these two functions actually handle a lot of the de-duplication heavy lifting for us.